plc/coqos_add.1
.TH coqos_add 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
coqos_add - Add a managed data stream

.SH SYNOPSIS
.BR coqos_add
.IR action
.IR priority 
.IR rate
.IR ttl
.IR operand
.IR condition 
.RI [ condition ]
[...]
.RI [ device ]
[...]

.SH DESCRIPTION
.B This page is under construction.

.PP
Add a managed stream to one or more powerline devices using the VS_CONN_ADD message.
Consult the Qualcomm Atheros Firmware Technical Reference Manual for a description of this vendor specific message type.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH BACKGROUND

.PP
Bandwidth management prioritizes streams so that data from lowest priority stream is dropped first whenever one of the following conditions is detected.
This ensures that the remaining streams continue operate at their full pontential.

.TP
.B Degraded line condition 
The channel degrades to the point where the available PHY rate from the transmitter to the receiver is too low due to the variability of the power line.s characteristics.

.TP
.B Over subscription 
Too much data is being sent per second, resulting in packet loss due to excessive collisions due to excessive channel oversubscription.

.TP
.B Lack of channel Capacity
On starting new source there is not enough channel capacity to support it.

.SH OPTIONS

.TP
.RB - e
Redirects stderr messages to stdout.
By convention progress messages and error messages are printed on stderr while primary program output is printed on stdout.
This option prints all output on stdout in cases where this is desired.

.TP
-\fBi \fIinterface\fR
Select the host Ethernet interface.
All requests are sent via this host interface and only reponses received via this host interface are recognized.
The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface.
This option then takes precedence over either default.

.TP
.RB - q
Suppresses status messages on stderr.

.TP
.RB - r
Display relative memory offsets on output.
This option is the default.

.TP
.RB - v
Prints additional information on stdout.
In particular, this option dumps outgoing Ethernet packets on stdout.

.TP
.RB - ? ,-- help
Displays program help information on stderr.
This option takes precedence over all other options on the command line except version information.

.TP
.RB - ! ,-- version
Displays program version information on stderr.
This option takes precedence over all other options on the command line except help information.
Use this option when sending screen dumps to Atheros technical staff.

.SH ARGUMENTS

.TP
.IR action
The action taken for frames that satisfy the selection criteria.
Valid actions are "\fBCAP0\fR", "\fBCAP1\fR", "\fBCAP2\fR", "\fBCAP3\fR" to specify the channel access priority queue.
CAP0 and CAP1 are for best effort data.
CAP2 is for video and non urgent MMEs.
CAP3 is for voice, urgent MMEs and control messages such as IGMP and MLD.

.TP
.IR priority
The relative priority of this stream.
Valid values are \fB0\fR through \fB15\fR where 0 is the lowest priority and 15 is the highest.

.TP
.IR destination
The destination MAC address.

.TP
.IR rate
The average expected data rate for this stream.
Valid values are \fB1\fR to \fB9000\fR where units are in 10kbps so the minimum rate is 10kbps and the maximum rate is 90mbps.

.TP
.IR ttl
The time to live for this stream.
Valid values are 10000 to 2000000 where units are in microseconds so the minimum is time is 10 milliseconds and the maximum time is 2 seconds.

.TP
.IR operand
The operand specifies the logical relationship between conditions before the \fIaction\fR to be taken.
Valid operands are listed and described under \fBOPERANDS\fR.

.TP
.IR condition
A conditional expression consisting of a \fIfield\fR, \fIoperator\fR and \fIvalue\fR.
See \fBCONDITIONS\fR for more information.

.TP
.IR device
The MAC address of some powerline device.
More than one address may be specified on the command line.
If more than one address is specified then operations are performed on each device in turn.
The default address is "\fBlocal\fR".
See \fBDEVICES\fR for more information.

.SH CONDITIONS
A condition consists of a \fIfield\fR, an \fIoperator\fR and a \fIvalue\fR.
One condition is required but three are permitted.
Condition order is not important but all conditions must appear after the \fIoperand\fR and before the \fIcontrol\fR.

.TP
.IR field
The field is the part of the Ethernet frame to be examined.
Some fields are not valid for some actions but this program does not enforce such rules since validation is performed by runtime firmware on each device.
Recognized fields are listed and described under \fBFIELDS\fR.

.TP
.IR operator
The operator specifies the relationsip that must exist between the \fIfield\fR and \fIvalue\fR in order for the \fIcondition\fR to evaluate \fBTrue\fR.
Currently, only equality operators are supported.
Valid operators are listed and described under \fBOPERATORS\fR.

.TP
.IR value
The value must be appropriate to the field type.
Some fields are MAC or IP addresses, some are integers, some are bitmaps and others are states.
Integers and bitmaps may be expressed in binary, decimal or hexadecimal format.
Binary values staRt with \fB0b\fR.
Hexadecimal values start with \fB0x\fR.
States are expressed using keywords.
Users are responsible for knowing how many bits are significant for each type of value.
Valid values are described along with fields under \fBFIELDS\fR.

.SH OPERANDS
The \fIoperand\fR indicates the logical relationship that must exist between conditions in the rule set before the action is applied to a frame.
Operands are expressed as discrete alphanumeric strings entered in upper, lower or mixed character case.
Failure to enter a known operand will result in an error message that lists all possible operands.
They are positon sensitive.
One operand is allowed and it must appear after the \fIaction\fR and before any \fIcondition\fR.

.TP
.BR Any
Apply the action to frames that satisfy \fBany\fR of the conditions.
This is equivalent to the logical \fBor\fR operation.

.TP
.BR All
Apply the action to frames that satisfy \fBall\fR of the conditions.
This is equivalent to the logical \fBand\fR operation.

.TP
.BR Always
Apply the action to all frames, regardless of any and all conditions that may be specified.

.SH FIELDS
Fields indicate the portion of the frame that is inspected during selection and the size and format of the value permited in the condition statement.
They are expressed as discrete alphanumeric strings entered in upper, lower or mixed character case.
Failure to enter a known field will result in an error message that lists all possible fields.

.TP
.BR ET
A 16-bit Ethertype expressed in decimal, hexadecimal or binary.
The format is described in IEEE Standard 802-2001 [4].

.TP 
.BR EthDA 
A 48-bit Ethernet destination address expressed in hexadecimal.
The format is described in IEEE Standard 802-2001 [4].

.TP 
.BR EthSA  
A 48-bit Ethernet source address expressed in hexadecimal.
The format is described in IEEE Standard 802-2001 [4].

.TP 
.BR VLANUP 
An 8-bit Ethernet VLAN tag where the lower 3 bits are the User Priority sub-field of a VLAN Tag defined in IEEE Std 802.1Q-1998 (Virtual Bridged Local Area Networks) [11].
The upper 5 bits should be zero.

.TP
.BR VLANID 
A 16-bit VLAN identifier where the lower 12 bits are the VLAN Identifier (VID) defined in IEEE Std 802.1Q-1998 (Virtual Bridged Local Area Networks) [11].
The upper 4 bits should be zero.

.TP
.BR IPV4TOS 
An 8-bit Type-of-Service code where the format is defined in the RFC 791 (Internet Protocol) [14].

.TP
.BR IPV4PROT 
An 8-bit Ethernet Protocol identifier.
The format is defined in the RFC 791 (Internet Protocol) [14].

.TP
.BR IPV4SA
A 32-bit Internet Protocol source address expressed in dotted-decimal notation.
The official format is defined in RFC 791 (Internet Protocol) [14].
Our implementation permits empty octets and leading zeros within fields.
For example, "..." is equivalent to "0.0.0.0 and "127..000.001" is equivalent to "127.0.0.1". 
.TP
.BR IPV4DA 
A 32-bit Internet Protocol destination address expressed in dotted-decimal notation. The official format is defined in RFC 791 (internet Protocol) [14]. Our implementation permits empty octets and leading zeros within fields. For example, "..." is equivalent to "0.0.0.0 and "127..000.001" is equivalent to "127.0.0.1".

.TP
.BR IPV6TC
An 8-bit Internet Protocol V6 traffic class expressed as defined in RFC 2460 (Internet Protocol Version 6) [17].

.TP
.BR IPV6FL
A 24-bit IPV6 flow label where the lower 20 bits are the IPv6 Flow Label defined in RFC 2460 (Internat Protocol Version 6) [17].
The upper 4 bits should be zero.
The value can be entered either as a decimal, binary or hex integer.

.TP
.BR IPV6SA
A 128-bit IPV6 source address expressed as colon-separated hexadecmial quartets (octet pairs).
The official format is defined in RFC 2460 (Internet Protocol Version 6) [17].
Our implementation permits multiple empty fields, abreviated fields and leading zeros within fields.
When multiple empty fields appear, only the right-most occurance expands to zeros.
For example, "FFFF::DDDD::BBBB::AAAA" is equivalent to "FFFF:0000:DDDD:0000:BBBB:0000:0000:AAAA".

.TP
.BR IPV6DA
A 128-bit IPV6 destination address expressed as colon-separated hexadecimal quartets (octet pairs).
The official format is defined in RFC 2460 (Internet Protocol Version 6) [17].
Our implementation permits multiple empty fields, abbreviated fields and leading zeros within fields.
When multiple empty fields appear, only the right-most occurance expands to zeros.
For example, "AAAA::BBBB::CCCC::DDDD" is equivalent to "AAAA:0000:BBBB:0000:CCCC:0000:0000:DDDD".

.TP
.BR TCPSP
A 16-bit TCP source port as a decimal integer.
The format is defined in RFC 793 (Transmission Control Protocol [15]).

.TP
.BR TCPDP 
A 16-bit TCP destination port expressed as a decimal integer.
The format is defined in RFC 793 (Transmission Control Protocol [15]).

.TP
.BR UDPSP 
A 16-bit UDP source port expressed as a decimal integer.
The format is defined in RFC 768 (User Datagram Protocol [13]).

.TP
.BR UDPDP
A 16-bit UDP destination port expressed as a decimal integer.
The format is defined in RFC 768 (User Datagram Protocol [13]).

.TP
.BR IPSP
A 16-bit IP source port expressed as a decimal integer.
This condition applies to either TCP or UDP packets, depending on the protocol used, and is valid only for actions "\fBCAP0\fR", "\fBCAP1\fR", "\fBCAP2\fR", "\fBCAP3\fR" and "\fBDrop\fR".

.TP
.BR IPDP 
A 16-bit IP destination port expressed as a decimal integer.
This condition applies to either TCP or UDP packets, depending on the protocol used, and is valid only for actions "\fBCAP0\fR", "\fBCAP1\fR", "\fBCAP2\fR", "\fBCAP3\fR" and "\fBDrop\fR".

.TP
.BR MME
A 24-bit Atheros HomePlugAV Management Message type expressed in hexadecimal.
The first byte is the MMV and the next two bytes are the MMTYPE for a HomePlug AV frame as defined in the HomePlug AV Specification.
The MMTYPE will match any MME sub-type (Request; Confirm; Indicate; Response).
This field is only valid for action "\fBBoost\fR".

.SH OPERATORS
An operator indicates an equality between a \fIfield\fR and a \fIvalue\fR.
An operator is an alphanumeric string entered in upper, lower or mixed character case.
Failure to enter a known operator will result in an error message that lists all possible operators.
Operators are position sensitive and must appear between each \fIfield\fR and \fIvalue\fR.

.TP
.BR Is
Indicates that the frame field must equal the associated value for the condition to evaluate true.

.TP
.BR Not
Indicates that the frame field must not equal the associated value for the condition to true true.

.SH STATES
A state is a special case of \fIvalue\fR.

.TP
.BR True , On , Yes , Present
Indicates a positive state or presence of some entity.
All are equivalent and can be used interchangeably.
Double negatives are permitted so "Is True" is equvalent to "Not False".

.TP
.BR False , Off , No , Missing
Indicates a negative state or absence of some entity.
ALl are equivalent and can be used interchangeably.
Double negatives are permitted so "Is False" is equvalent to "Not True".

.SH DEVICES
Powerline devices use Ethernet Media Access Control (MAC) addresses.
A MAC address is a 48-bit value entered as 12 hexadecimal digits in upper, lower or mixed character case.
Octets may be separated with colons for clarity.
For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.

.PP
The following MAC addresses are common or special and may be entered by name instead of number.

.TP
.BR all
Same as "broadcast".

.TP
.BR broadcast
A synonym for the Ethernet broadcast address, \fBFF:FF:FF:FF:FF:FF\fR.
All devices, whether local, remote or foreign recognize messages sent to this address.
A remote device is any device at the far end of a powerline connection.
A foreign device is any device not manufactured by Atheros.

.TP
.BR local
A synonym for the Qualcomm Atheros vendor specific Local Management Address (LMA), \fB00:B0:52:00:00:01\fR.
All local Atheros devices recognize this address but remote and foreign devices do not.
A remote device is any device at the far end of a powerline connection.
A foreign device is any device not manufactured by Atheros.

.SH REFERENCES
See the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual for more information.

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information may not be available.
Qualcomm Atheros reserves the right to modify message structure and content in future firmware releases without any obligation to notify or compensate users of this program.

.SH EXAMPLES
The following example adds a temporary stream to device \fB00:b0:52:BA:BE:01\fR which will then manage the bandwidth for that stream until removed with program \fBcoqos_rel\fR.

.PP
	# coqos_add CAP2 15 5000 200000 any ethda is 192.168.105 00:B0:52:BA:BE:01         

.PP
This adds a stream to the bandwidth manager that sets all traffic to destination address of 192.168.0.105 as priority 15 (the highest priority).
This would then need to also be sent to each device on the network.
Refer to int6krule for more details on how to specify conditions.

.SH SEE ALSO
.BR coqos_info ( 1 ),
.BR coqos_man ( 1 ),
.BR coqos_mod ( 1 ),
.BR coqos_rel ( 1 )

.SH CREDITS
 Bill Wike
 Nathaniel Houghton
 Charles Maier



CHANGES
open-plc-utils-0.0.6
---------------

1. Programs mac2pw and mac2pwd implement a new method of generating device passwords. The old method is supported but is not the default. 

open-plc-utils-0.0.5
---------------

open-plc-utils-0.0.4
---------------

open-plc-utils-0.0.3
---------------

1. Added header VisualStudioNET/include/inttype.h to define print specifier constant PRId64 for Windows systems.

2. Replaced conditional compiler code around print specifiers "lld" and "I64d" and replaced single specifier PRId64 as defined by the system.

3. Renamed man pages ending in .7 with .1 to reflect that fact that they will now be installed man1 instead of man7.

4. Replaced const HOMEPLUG_MTYPE with constant ETH_P_HPAV throughout.

5. Correct endianess of FIELD_ETH_TYPE in module coqos_add.c.

6. Updated chipset codes and implemented table-driven chipset detection algorithm.

7. Moved TLVNode structure definition from nvm/nvm.h to tools/tlv.h to prepare for exanded use of TLV structures in future MMEs.


open-plc-utils-0.0.2
---------------

1. Removed references to private vendor-specific MMEs

2. Removed unused archive targets in primary Makefile.

3. Added .gitignore to subdirectories where programs and other objects are compiled.

4. Added conditional __UCLIBC__ compile statement to ARCPPrint.c to account for variable argument definition error during compile; 

5. Added 'ignore' target to cascading Makefiles to re-write local .gitignore files.

6. Corrected email address for Florian Fainelli in module ARPCPrint.

7. Added new HomePlug AV SLAC Protocol MMEs definitions to homeplug.h.

8. Modified makefile 'ignore' target to use tr instead of sed because some sed implementations do not accept '\n'.



plc/coqos_add.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted (subject to the limitations
 *   in the disclaimer below) provided that the following conditions
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of
 *     its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   coqos_add.c - Add MCMSA Stream
 *
 *
 *   Contributor(s):
 *	Bill Wike
 *	Nathaniel Houghton
 *	Charles Maier
 *
 *--------------------------------------------------------------------*/


/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../plc/plc.h"
#include "../plc/coqos.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/basespec.c"
#include "../tools/uintspec.c"
#include "../tools/hexdump.c"
#include "../tools/hexencode.c"
#include "../tools/bytespec.c"
#include "../tools/ipv4spec.c"
#include "../tools/ipv6spec.c"
#include "../tools/hexdecode.c"
#include "../tools/todigit.c"
#include "../tools/typename.c"
#include "../tools/error.c"
#include "../tools/synonym.c"
#include "../tools/lookup.c"
#include "../tools/assist.c"
#include "../tools/codelist.c"
#endif

#ifndef MAKEFILE
#include "../plc/Request.c"
#include "../plc/Confirm.c"
#include "../plc/Failure.c"
#include "../plc/Display.c"
#include "../plc/ReadMME.c"
#include "../plc/SendMME.c"
#include "../plc/Devices.c"
#include "../plc/rules.c"
#endif

#ifndef MAKEFILE
#include "../ether/channel.c"
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#endif

#ifndef MAKEFILE
#include "../mme/MMECode.c"
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/UnwantedMessage.c"
#endif


/*====================================================================*
 *
 *   signed add_conn (struct plc * plc, struct connection * connection);
 *
 *
 *
 *   Contributor(s):
 *	Bill Wike
 *
 *--------------------------------------------------------------------*/

static signed add_conn (struct plc * plc, struct connection * connection)

{
	struct channel * channel = (struct channel *)(plc->channel);
	struct message * message = (struct message *)(plc->message);

#ifndef __GNUC__
#pragma pack (push,1)
#endif

	struct __packed vs_add_conn_req
	{
		struct ethernet_hdr ethernet;
		struct qualcomm_hdr qualcomm;
		struct connection connection;
	}
	* request = (struct vs_add_conn_req *)(message);
	struct __packed vs_add_conn_cnf
	{
		struct ethernet_hdr ethernet;
		struct qualcomm_hdr qualcomm;
		uint32_t REQ_ID;
		uint8_t MSTATUS;
		uint16_t ERR_REC_CODE;
		uint32_t RSVD;
		uint16_t CID;
		uint8_t REJECT_MAC [ETHER_ADDR_LEN];
		uint16_t CSPEC_VERSION;
		uint8_t CONN_CAP;
		uint8_t CONN_COQOS_PRIO;
		uint16_t CONN_RATE;
		uint32_t CONN_TTL;
	}
	* confirm = (struct vs_add_conn_cnf *)(message);

#ifndef __GNUC__
#pragma pack (pop)
#endif

	Request (plc, "Add COQOS connection");
	memset (message, 0, sizeof (* message));
	EthernetHeader (&request->ethernet, channel->peer, channel->host, channel->type);
	QualcommHeader (&request->qualcomm, 0, (VS_CONN_ADD | MMTYPE_REQ));
	memcpy (&request->connection, connection, sizeof (struct connection));
	memcpy (&request->connection.rule.CLASSIFIERS [request->connection.rule.NUM_CLASSIFIERS], &request->connection.cspec, sizeof (request->connection.cspec));
	plc->packetsize = sizeof (struct vs_add_conn_req);
	if (SendMME (plc) <= 0)
	{
		error (PLC_EXIT (plc), errno, CHANNEL_CANTSEND);
		return (-1);
	}
	while (ReadMME (plc, 0, (VS_CONN_ADD | MMTYPE_CNF)) <= 0)
	{
		if (confirm->MSTATUS)
		{
			Failure (plc, "Device won't add connection");
			return (-1);
		}
		Confirm (plc, "Setting ...");
	}
	return (0);
}


/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"ei:qv",
		"action priority destination rate ttl operand condition [...] [device] [...]\n\n          where condition is field operator value",
		"CoQos Stream Utility",
		"e\tredirect stderr to stdout",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"q\tquiet mode",
		"v\tverbose mode",
		(char const *) (0)
	};

#include "../plc/plc.c"

	struct connection connection;
	struct MMEClassifier * rule = (struct MMEClassifier *)(&connection.rule.CLASSIFIERS);
	uint16_t * word;
	uint8_t * byte;
	signed code;
	signed c;
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			_setbits (plc.flags, PLC_SILENCE);
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (plc.flags, PLC_VERBOSE);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	memset (&connection, 0, sizeof (connection));
	if ((code = lookup (* argv++, actions, SIZEOF (actions))) == -1)
	{
		assist (*--argv, CLASSIFIER_ACTION_NAME, actions, SIZEOF (actions));
	}
	connection.cspec.CONN_CAP = (uint8_t)(code);
	argc--;
	if (!argc)
	{
		error (1, ECANCELED, "Expected Priority: 0-15");
	}
	connection.cspec.CONN_COQOS_PRIO = (uint8_t)(uintspec (* argv++, 0, 15));
	argc--;
	if (!argc)
	{
		error (1, ECANCELED, "Expected Destination MAC Address");
	}
	if (!hexencode (connection.APP_DA, sizeof (connection.APP_DA), synonym (* argv++, devices, SIZEOF (devices))))
	{
		error (1, errno, "Invalid MAC=[%s]", *--argv);
	}
	argc--;
	if (!argc)
	{
		error (1, ECANCELED, "Expected Data Rate: 10-9000 (kbps)");
	}
	connection.cspec.CONN_RATE = (uint16_t)(uintspec (* argv++, 1, 9000));
	argc--;
	if (!argc)
	{
		error (1, ECANCELED, "Expected TTL: 10000-2000000 (microseconds)");
	}
	connection.cspec.CONN_TTL = (uint32_t)(uintspec (* argv++, 10000, 2000000));
	argc--;
	if ((code = lookup (* argv++, operands, SIZEOF (operands))) == -1)
	{
		assist (*--argv, CLASSIFIER_OPERAND_NAME, operands, SIZEOF (operands));
	}
	connection.rule.MOPERAND = (uint8_t)(code);
	argc--;
	while ((* argv) && (lookup (* argv, controls, SIZEOF (controls)) == -1))
	{
		if ((code = lookup (* argv++, fields, SIZEOF (fields))) == -1)
		{
			assist (*--argv, CLASSIFIER_FIELD_NAME, fields, SIZEOF (fields));
		}
		rule->CR_PID = (uint8_t)(code);
		argc--;
		if ((code = lookup (* argv++, operators, SIZEOF (operators))) == -1)
		{
			assist (*--argv, CLASSIFIER_OPERATOR_NAME, operators, SIZEOF (operators));
		}
		rule->CR_OPERAND = (uint8_t)(code);
		argc--;
		if (!argc || !* argv)
		{
			error (1, ENOTSUP, "Have %s '%s' without any value", CLASSIFIER_OPERATOR_NAME, *--argv);
		}
		switch (rule->CR_PID)
		{
		case FIELD_ETH_SA:
		case FIELD_ETH_DA:
			bytespec (* argv++, rule->CR_VALUE, ETHER_ADDR_LEN);
			break;
		case FIELD_IPV4_SA:
		case FIELD_IPV4_DA:
			ipv4spec (* argv++, rule->CR_VALUE);
			break;
		case FIELD_IPV6_SA:
		case FIELD_IPV6_DA:
			ipv6spec (* argv++, rule->CR_VALUE);
			break;
		case FIELD_VLAN_UP:
		case FIELD_IPV4_TOS:
		case FIELD_IPV4_PROT:
			byte = (uint8_t *)(rule->CR_VALUE);
			*byte = (uint8_t)(basespec (* argv++, 0, sizeof (* byte)));
			break;
		case FIELD_VLAN_ID:
		case FIELD_TCP_SP:
		case FIELD_TCP_DP:
		case FIELD_UDP_SP:
		case FIELD_UDP_DP:
		case FIELD_IP_SP:
		case FIELD_IP_DP:
			word = (uint16_t *)(rule->CR_VALUE);
			*word = (uint16_t)(basespec (* argv++, 0, sizeof (* word)));
			*word = htons (*word);
			break;
		case FIELD_ETH_TYPE:
			word = (uint16_t *)(rule->CR_VALUE);
			*word = (uint16_t)(basespec (* argv++, 0, sizeof (* word)));
			*word = htons (*word);
			break;
		case FIELD_HPAV_MME:
			bytespec (* argv++, rule->CR_VALUE, sizeof (uint8_t) + sizeof (uint16_t));
			byte = (uint8_t *)(rule->CR_VALUE);
			HTOBE16 ((uint16_t)(* ++byte));
			break;
		case FIELD_IPV6_TC:
		case FIELD_IPV6_FL:
		case FIELD_TCP_ACK:
		default:
			error (1, ENOTSUP, "Field '%s' (0x%02X)", argv [-2], rule->CR_PID);
			break;
		}
		connection.rule.NUM_CLASSIFIERS++;
		if (connection.rule.NUM_CLASSIFIERS > RULE_MAX_CLASSIFIERS)
		{
			error (1, ENOTSUP, "More than %d classifiers in rule", RULE_MAX_CLASSIFIERS);
		}
		rule++;
		argc--;
	}
	connection.cspec.CSPEC_VERSION = 0x0001;
	openchannel (&channel);
	if (!(plc.message = malloc (sizeof (* plc.message))))
	{
		error (1, errno, PLC_NOMEMORY);
	}
	if (!argc)
	{
		add_conn (&plc, &connection);
	}
	while ((argc) && (* argv))
	{
		if (!hexencode (channel.peer, sizeof (channel.peer), synonym (* argv, devices, SIZEOF (devices))))
		{
			error (1, errno, PLC_BAD_MAC, * argv);
		}
		add_conn (&plc, &connection);
		argc--;
		argv++;
	}
	free (plc.message);
	closechannel (&channel);
	exit (0);
}

